rootston/cursor.c

```diff
@@ -76,27 +78,153 @@ static void cursor_set_xcursor_image(struct roots_input *input,
+			wl_resource_get_client(layer_surface->resource);
+		struct wlr_seat_handle *handle =
+			wlr_seat_handle_for_client(input->wl_seat, client);
+
+		double sx, sy;
+		bool ok = layer_surface_is_at(input->server->desktop, layer_surface,
+			x, y, &sx, &sy);
+		if (!ok) {
+			if (focused_layer_surface != NULL &&
+					focused_layer_surface->layer_surface == layer_surface) {
+				// We're leaving this surface
+				*focused_list = (*focused_list)->prev;
+				wl_list_remove(&focused_layer_surface->link);
+				free(focused_layer_surface);
+
+				if (handle && handle->pointer) {
+					uint32_t serial =
+						wl_display_next_serial(handle->wlr_seat->display);
+					wl_pointer_send_leave(handle->pointer, serial,
+						layer_surface->surface->resource);
+				}
+			}
+			continue;
+		}
+
+		if (layer_surface->current->exclusive_types &
+				LAYER_SURFACE_INPUT_DEVICE_POINTER) {
+			*exclusive = layer_surface;
+			*exclusive_sx = sx;
+			*exclusive_sy = sy;
+			return NULL;
+		}
+
+		if (focused_layer_surface &&
+				focused_layer_surface->layer_surface == layer_surface) {
+			// This surface was already focused last time
+			*focused_list = (*focused_list)->prev;
+		} else if (!focused_layer_surface ||
+				focused_layer_surface->layer_surface != layer_surface) {
+			// We're entering this surface
+			focused_layer_surface =
+				calloc(1, sizeof(struct roots_focused_layer_surface));
+			if (focused_layer_surface == NULL) {
+				return NULL;
+			}
+			focused_layer_surface->layer_surface = layer_surface;
+			wl_list_insert(*focused_list, &focused_layer_surface->link);
+
+			if (handle && handle->pointer) {
+				uint32_t serial =
+					wl_display_next_serial(handle->wlr_seat->display);
+				wl_pointer_send_enter(handle->pointer, serial,
+					layer_surface->surface->resource, wl_fixed_from_double(sx),
+					wl_fixed_from_double(sy));
+			}
+		}
+
+		if (handle && handle->pointer) {
+			wl_pointer_send_motion(handle->pointer, time,
+				wl_fixed_from_double(sx), wl_fixed_from_double(sy));
+		}
+	}
+
+	return NULL;
+}
+
 void cursor_update_position(struct roots_input *input, uint32_t time) {
 	struct roots_desktop *desktop = input->server->desktop;
-	struct roots_view *view;
-	struct wlr_surface *surface;
+	struct wlr_surface *surface = NULL;
 	double sx, sy;
+	struct wlr_layer_surface *layer_surface = NULL;
+	struct wl_list *remaining_layer_surfaces;
+	struct wl_list *focused_layer_surfaces;
 	switch (input->mode) {
 	case ROOTS_CURSOR_PASSTHROUGH:
-		view = view_at(desktop, input->cursor->x, input->cursor->y,
-			&surface, &sx, &sy);
-		bool set_compositor_cursor = !view && input->cursor_client;
-		if (view) {
-			struct wl_client *view_client =
-				wl_resource_get_client(view->wlr_surface->resource);
-			set_compositor_cursor = view_client != input->cursor_client;
+		// Send events to non-exclusive layer surfaces, check if a layer surface
+		// gets exclusive events
+		remaining_layer_surfaces = &desktop->surface_layers->surfaces;
```

Comment:
You really have to drill "layer surface" into your head to get this straight. `surface_layers` really sounds like it should be a list of layer surfaces. What about calling the global the "layer surface shell"?

Reply:
To clarify the naming convention I had in mind when I originally wrote the protocol: *surface layers* refer to the layers in which surfaces may reside and the protocol as a whole. Layer surfaces is just short for "surface layer surface". I'm not torn up about how we name anything, just clarifying.

Reply:
> surface layers refer to the layers in which surfaces may reside and the protocol as a whole

So we have one term with three meanings. The protocol, an object in the protocol, and an attribute of an object in the protocol.

Reply:
(@SirCmpwn edited his reply)

So yeah basically I'm trying to use terms consistently:
* _Surface layers_ for the protocol
* _Layer surface_ for a surface
* _Layer surfaces_ for multiple surfaces (such as in a list of surfaces)